File: leetcode/python/0088_merge_sorted_array.py
class Solution(object):
    def merge(self, nums1, m, nums2, n):
        """
        :type nums1: List[int]
        :type m: int
        :type nums2: List[int]
        :type n: int
        :rtype: None Do not return anything, modify nums1 in-place instead.
        """
        # we can work backwards from m to 0 for nums1, n to 0 for nums2 inserting at m + n towards 0
        # i in range(m, -1, -1), j in range (n, -1, -1), k in range (m + n, -1, -1), while k >= 0:
        # -> if nums1[i] < nums2[j]: nums1[k] = nums2[j], j--, k--
        # -> elif nums1[i] >= nums2[j]: nums1[k] = nums1[i], i--, k--
        # return nums1

        if n == 0: return

        i, j, k = m - 1, n - 1, m + n - 1

        while j >= 0:
            if i >= 0 and nums1[i] > nums2[j]:
                nums1[k] = nums1[i]
                i -= 1
                k -= 1
            else:
                nums1[k] = nums2[j]
                j -= 1
                k -= 1
        

File: leetcode/cpp/0088_merge_sorted_array.cpp
class Solution {
public:
    void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
        int sizeM = m;

        for (int i = 0; i < n; i++) {
            if (m == 0) {
                nums1.insert(nums1.begin(), nums2.begin(), nums2.end());
                sizeM++;
                break;
            }

            if (nums2[i] <= nums1[0]) {
                nums1.insert(nums1.begin(), nums2[i]);
                sizeM++;
                continue;
            }

            for (int j = 0; j < nums1.size(); j++) {
                if (nums1[j] < nums2[i] && nums2[i] <= nums1[j + 1]) {
                    nums1.insert(nums1.begin() + j + 1, nums2[i]);
                    sizeM++;
                    break;
                }

                if (j == sizeM - 1 && nums1[j] < nums2[i]) {
                    nums1.insert(nums1.begin() + sizeM, nums2[i]);
                    sizeM++;
                    break;
                }
            }
        }

        nums1.erase(nums1.begin() + m + n, nums1.begin() + nums1.size());
    }
};